Add initial mkdocs site (#1088)
google/fhir-data-pipes: doc/docs/concepts.md

2. **Flexible and scalable deployment approaches**: Deploy on a single machine or multi-worker distributed environments
3. **Simplify querying:** Multiple approaches to flatten FHIR resources to make developing analytics solutions easier

## FHIR Data Pipes
FHIR Data Pipes is built on [Apache Beam](https://beam.apache.org/) technology **designed for ease of scalability** and has multiple deployment options from _local, on-prem to cloud_.

FHIR Data Pipes is made up of the **ETL Pipeline** and **Controller** modules **that are designed to work together** to provide continuous transformation of FHIR data to Apache Parquet files for data analysis or another FHIR server for data integration. 

![FHIR Data Pipes Transform Step Image](images/ETL_FHIR_to_Parquet.png)

## ETL Pipelines
Java JAR to **Extract**, **Transform** and **Load** FHIR data from a FHIR input to FHIR-in-Parquet output.

### Extraction

FHIR Data Pipes is designed to work with FHIR source data in various forms: 

* Valid FHIR Server API (e.g. HAPI FHIR)
* FHIR facade implementations (e.g. OpenMRSv3)
* Bulk Export API data
* ndjson representation

### Transformation

> FHIR Resources are transformed into a "Parquet on FHIR" format:

Author: I thought we settled on "Parquet in FHIR"?

> FHIR Resources are transformed into a "Parquet on FHIR" format:

Reviewer: I believe "FHIR in Parquet" is what we settled on which I think you have also applied on the initial image in the intro page: https://fredhersch.github.io/openmrs-fhir-analytics/

> FHIR Resources are transformed into a "Parquet on FHIR" format:

Reviewer: I'll address this in a future PR too (resolving all comments).

* The Pipelines Controller is built on top of pipelines and shares many of the same settings
* Using the controller module you can schedule periodic incremental updates or use a [web control panel](link_to_section_in_adv_guide) to start the pipeline manually

## Flexible and scalable deployment approaches

There are a number of different deployment approaches - see table below. 

Choosing the right approach, comes down to evaluation of a number of factors including the specific use case, requirements of the project, future scalability needs and capabilities of the team. 

| Scenario | Approach | Considerations |
| -------- | ----------| -------------- |
|  Simple relational database to power dashboards or reporting | Custom schema defined as ViewDefinition Resources - [see tutorial](tutorial_lossy_example/) | By design this will provide a constrained set of variables in the views |
| Need for a horizontally scalable architecture | Parquet based DWH and distributed query engine (e.g. SparkSQL) - [see tutorial](tutorial_add_dashboard/) | Will need to manage both distributed storage (Parquet files) and a distributed query engine |

> | Parquet as DWH with a non-distributed OLAP database engine | Parquet based DWH with non-distributed query engine set-up (e.g. SparkSQL or DuckDB) | Scalability and performance (e.g. SparkSQL >> DuckDB) |

Reviewer: ```suggestion
| Lossless FHIR DWH with a single process query engine | Parquet based DWH with non-distributed query engine set-up (e.g. single-node Spark or DuckDB) | Scalability and performance (e.g. SparkSQL >> DuckDB) |
```

> | Exploratory data science or ML use cases | Use the generated Parquet files which as _"near lossless"_ for enhanced data science workflows | Can either use the Parquet or custom schema to power dashboards or reports |

Reviewer: We can probably delete this row; I am not sure if adds value.

Author: I did think it was worth making this explicit. Ok to remove it if you think its superfluous or obvious

Reviewer: Okay sure.

> | Push FHIR data to a central FHIR Data Store for a Shared Health Record or single repository for analytics workloads | Use the Pipelines Controller to push from a FHIR source to a FHIR sink | Management of the intermediate Parquet files created as part of the pipelines | 

Reviewer: ```suggestion
| Push FHIR data to a central FHIR-store (e.g., for a Shared Health Record system) | Use the Pipelines Controller to push from a FHIR source to a FHIR sink | Management of the intermediate Parquet files created as part of the pipelines | 
```

> ## Simplify querying with Pre-defined views

Reviewer: ```suggestion
## Query simplification approaches with pre-defined views
```

The heavily nested nature of FHIR resources requires complex SQL queries that can make it difficult to work with for analytics use cases. A common approach to address this is to flatten the data into a set of views (virtual or materialized) which can then be queried using simpler SQL statements. 

> FHIR Data Pipes provides two approaches for flattening the FHIR Resources into virtual or materialized views

Reviewer: ```suggestion
FHIR Data Pipes provides two approaches for flattening the FHIR resources into virtual or materialized views:
```

1.  SQL queries to generate virtual views (outside the pipeline)

> 2.  FHIR ViewDefinition resources to generated materialized views (within the pipeline)

Reviewer: ```suggestion
2.  FHIR [ViewDefinition](https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/StructureDefinition-ViewDefinition.html)
resources to generate materialized views (within the pipeline)
```

> For both of these approaches, a set of **"predefined views”** for common resources are provided. These can be modified or extended.

Reviewer: ```suggestion
For both of these approaches, a set of [**"predefined views”**](https://github.com/google/fhir-data-pipes/tree/master/docker/config/views) for common resources are provided. These can be modified or extended.
```

> The currently supported list (as at 1st June, 2024), which can be found in the `docker/config/views' directory are:

Reviewer: ```suggestion
The currently supported list (as of June, 2024) are:
```

```

> Condition, DiagnosticReport, Encounter, Immunization, Location, Medicationrequest, Observation, Organization, Patient, Practitioner, PractitionerRole, Procedure

Reviewer: It is also probably better to remove "```" and make this an itemized list with links to FHIR resource pages for each.
```suggestion
Condition, DiagnosticReport, Encounter, Immunization, Location, Medicationrequest, Observation, Organization, Patient, Practitioner, PractitionerRole, Procedure
```
```suggestion
Condition
DiagnosticReport
Encounter
Immunization
Location
Medicationrequest
Observation
Organization
Patient
Practitioner
PractitionerRole
Procedure
```

```

### SQL virtual views

> These are samples of more complex SQL-on-FHIR queries for defining flat views for common FHIR Resources. These virtual views are applied outside of the pipelines in the downstream SQL query engine. 

Reviewer: ```suggestion
These are samples of more complex SQL-on-FHIR queries for defining flat views for common FHIR resources. These virtual views are applied outside of the pipelines in the downstream SQL query engine. 
```

> The queries, which have .sql suffix can be found here: /docker/config/views (e.g Patient_flat.sql)

Reviewer: ```suggestion
The queries, which have `.sql` suffix, can be found in [/docker/config/views](https://github.com/google/fhir-data-pipes/tree/master/docker/config/views) directory (e.g `Patient_flat.sql`).
```

> An example of a flat view for the Observation Resource is below

Reviewer: ```suggestion
An example of a flat view for the Observation resource is below:
```

> ```sql

Reviewer: Please copy the example from [here](https://github.com/google/fhir-data-pipes/blob/master/docker/config/views/Observation_flat.sql) instead.

Reviewer: I'll address this in a future PR.

CREATE OR REPLACE VIEW flat_observation AS
SELECT O.id AS obs_id, O.subject.PatientId AS patient_id,
        OCC.`system` AS code_sys, OCC.code,
        O.value.quantity.value AS val_quantity,
        OVCC.code AS val_code, OVCC.`system` AS val_sys,
        O.effective.dateTime AS obs_date
      FROM Observation AS O LATERAL VIEW OUTER explode(code.coding) AS OCC
        LATERAL VIEW OUTER explode(O.value.codeableConcept.coding) AS OVCC
```

### ViewDefinition resource

> The [SQL-on-FHIR-v2 specification](https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/) defines a standards based pattern for defining Views as FHIRPath expressions in a logical structure to specify the column names and values (as unnested items).

Reviewer: ```suggestion
The [SQL-on-FHIR-v2 specification](https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/) defines a ViewDefinition resource for defining views. Each column in the view is defined using a FHIRPath expression. There is also an unnesting construct and support for `constant` and `where` clauses too.
```

A system (pipeline or library) that implements the “View Layer” of the specification provides a View Runner that is able to process these FHIR ViewDefinition Resources over the “Data Layer” (lossless representation of the FHIR data). The output of this are a set of portable, tabular views that can be consumed by the “Analytics Layer” which is any number of tools that can be used to work with the resulting tabular data.
 

> FHIR Data Pipes is a reference implementation of the SQL-on-FHIR-v2 specification:

Reviewer: ```suggestion
FHIR Data Pipes is a [reference implementation](https://fhir.github.io/sql-on-fhir-v2/#impls) of the SQL-on-FHIR-v2 specification:
```

*   The "View Runner" is by default part of the ETL Pipelines and uses the transformed Parquet files as the “Data Layer”. _This can be extracted to be a stand-alone component if required_

> *   When enabled as part of the Pipeline configuration, it will apply the ViewDefinition Resources from the /config/views folder and materialize the resulting tables to the PostgresSQL database.

Reviewer: ```suggestion
*   When enabled as part of the Pipeline configuration, it will apply the ViewDefinition resources from the [views folder](https://github.com/google/fhir-data-pipes/tree/master/docker/config/views) and materialize the resulting tables to the configured database (e.g., an instance of PostgresSQL, MySQL, etc.).
```

> *   A set of pre-defined ViewDefinitions for common FHIR Resources is provided as part of the default package. These can be adapted, replaced and extended

Reviewer: ```suggestion
*   A set of pre-defined ViewDefinitions for common FHIR resources is provided as part of the default package. These can be adapted, replaced and extended.
```

> *   The FHIR Data Pipes provides a simple ViewDefinition Editor which can be used to explore FHIR ViewDefinitions and apply these to individual FHIR Resources

Reviewer: ```suggestion
*   The FHIR Data Pipes provides a simple ViewDefinition Editor which can be used to explore FHIR ViewDefinitions and apply these to individual FHIR resources.
```

Once the FHIR data has been transformed via the ETL Pipelines, the resulting schema is available for querying using a JDBC interface. 

### ViewDefinition editor
The ViewDefinition editor provides a way to quickly evaluate ViewDefinition resources against sample FHIR data. You access it as part of the Web Control Panel, selecting the "Views" navigation item in the top right corner.

Using the ViewDefinition editor you can:

* Provide an input ViewDefinition (left)

> * Apply it to a sample input FHIR Resource (right pane)

Reviewer: ```suggestion
* Apply it to a sample input FHIR resource (right pane)
```
